### DISCUSSION TITLE
Support copy a file to multiple folders

### DISCUSSION THREAD

**Discussion by mingzhao2019** [doublecmd/doublecmd#1699] [Suggestions]
FYI, in totalcmd, it is possible to copy one file to multiple folders simultaneously, e.g.,

1. select one file in the left panel
2. select multiple folders in the right panel
3. in the left panel, press F5, choose options, and select 'Copy to all selected folders or shortcuts in the target panel (P)'

Is it possible for doublecmd to support this feature?

**Comment by ags1234** [comment]
This would be very useful to have. Are there any dual panel file managers which do this?

**Reply by mingzhao2019** [reply]
yes, totalcmd does

**Comment by Skif-off** [comment]
If the matter is urgent, then automation is possible (in theory), for example: [get a list of tabs](https://github.com/j2969719/doublecmd-plugins/blob/master/utils/ChooseTab.lua) + copying & queue ([1](https://github.com/j2969719/doublecmd-plugins/blob/master/scripts/lua/copytoallsubdirs.lua) and/or [2](https://github.com/j2969719/doublecmd-plugins/blob/master/scripts/lua/copytosubdirs.lua)).

**Reply by mingzhao2019** [reply]
could you please explain more in detail? how to do these steps for example? I tried use the lua file but seems cannot combine with copying & queue

**Reply by Skif-off** [reply]
I've never used `copyto*.lua `, I just know they exist (I think I've seen discussions about these scripts, but I couldn't find it).

Also
```lua
local Target = Directory:gsub(' ', '\\ ');
```
tells us that this is not a cross-platform solution.

Give me some time, I'll think about how to put all this together to solve the task.

**Reply by Skif-off** [reply]
Maybe something like this
```lua
-- CopyToTargetSelectedFolders.lua (cross-platform)
-- 2024.06.04
--[[
Copy selected files to all selected folders in the target directory.

The script uses the navigation commands and "cm_Copy" with "confirmation=" & "queueid=1"
(see details here https://doublecmd.github.io/doc/en/cmds.html#cm_Copy).

How to use: add button (https://doublecmd.github.io/doc/en/toolbar.html) with
internal command: cm_ExecuteScript
params:
  path-to-script.lua
  %"0%LUt
  %"0%Dt
  confirmation

"confirmation" is optional parameter: copy operation will begin with a confirmation window
(https://doublecmd.github.io/doc/en/copymove.html#confirmation) for each folder.
]]

local params = {...}

local sScrName = debug.getinfo(1).source
if string.sub(sScrName, 1, 1) == "@" then
  sScrName = string.sub(sScrName, 2, -1)
end
sScrName = SysUtils.ExtractFileName(sScrName)

if #params < 2 then
  Dialogs.MessageBox("Check the list of parameters!", sScrName, 0x0030)
  return
end

local aDirs  = {}
local sConfirmation = "0"
if (#params == 3) and (params[3] == "confirmation") then
  sConfirmation = "1"
end

local function GetTargetDirs(p)
  local h, err = io.open(p, "r")
  if h == nil then
    Dialogs.MessageBox("Error: " .. err, sScrName, 0x0030)
    return nil
  end
  local i
  for l in h:lines() do
    i = SysUtils.FileGetAttr(l)
    if math.floor(i / 0x00000010) % 2 ~= 0 then
      table.insert(aDirs, l)
    end
  end
  h:close()
end

local function QuoteIt(s)
  if string.find(s, " ", 1, true) == nil then return s end
  if SysUtils.PathDelim == "\\" then
    return '"' .. s .. '"'
  else
    return string.gsub(s, " ", "\\ ")
  end
end

local function CDAndCopy(sDir)
  DC.ExecuteCommand("cm_FocusSwap")
  DC.ExecuteCommand("cm_ChangeDir", QuoteIt(sDir))
  DC.ExecuteCommand("cm_FocusSwap")
  DC.ExecuteCommand("cm_SaveSelection")
  DC.ExecuteCommand("cm_Copy", "confirmation=" .. sConfirmation, "queueid=1")
  DC.ExecuteCommand("cm_RestoreSelection")
end

GetTargetDirs(params[1])
if #aDirs == 0 then
  Dialogs.MessageBox("Nothing selected?", sScrName, 0x0030)
  return
end

local sTmp
if #aDirs > 3 then
  sTmp = table.concat(aDirs, "\n", 1, 3) .. "\n..."
else
  sTmp = table.concat(aDirs, "\n")
end
local sMsg = "Copy to all selected folders (" .. #aDirs .. ") in the target directory?\n\n" .. sTmp
local iRet = Dialogs.MessageBox(sMsg, sScrName, 0x0001 + 0x0020)
if iRet == 0x0001 then
  for i = 1, #aDirs do
    CDAndCopy(aDirs[i])
  end
  DC.ExecuteCommand("cm_FocusSwap")
  DC.ExecuteCommand("cm_ChangeDir", QuoteIt(params[2]))
  DC.ExecuteCommand("cm_FocusSwap")
end
```
See the description in the beginning of the script and [this](https://doublecmd.github.io/doc/en/lua.html#dllrequired) subsection.

**Reply by mingzhao2019** [reply]
it works for me, thanks so much!

**Comment by rmdug** [comment]
Hello,

I would also be very interested by such a feature inside Double Commander.
If the copy was able to follow symlinks or a list of paths it would also be great !!

Regards,

Rémi

**Reply by mingzhao2019** [reply]
hello, perhaps you can try the lua script above-mentioned, it works for me in local environment.

**Reply by rmdug** [reply]
Hello Mingzhao and skif-off

Thanks for your proposal : )
I just tried it. It works very well on networks folders, but unfortunately it is unable to follow symlinks.
In fact I need to dispatch a file into several folders that are scattered in different places, so it is not really useful for me to copy it inside several folders in a unique target directory.

I was thinking of those turnarounds:
- As mentioned in this thread TC seems to be able to perform this task: [https://www.ghisler.ch/board/viewtopic.php?t=39156](https://www.ghisler.ch/board/viewtopic.php?t=39156)
- Copy the file with a script similar to the one proposed by skif-off, but with the ability to follow symlinks;
- Copy the file from the left panel to all right tabs directories.

Do you think that this would be achievable ?

Thanks a lot for your attention !

Regards,

Rémi

**Reply by Skif-off** [reply]
@rmdug 

> unfortunately it is unable to follow symlinks

Any symlinks? Do you mean symlinks in the target directory? I checked in Windows, they open normally.

**Reply by mingzhao2019** [reply]
> Hello Mingzhao and skif-off
> 
> Thanks for your proposal : ) I just tried it. It works very well on networks folders, but unfortunately it is unable to follow symlinks. In fact I need to dispatch a file into several folders that are scattered in different places, so it is not really useful for me to copy it inside several folders in a unique target directory.
> 
> I was thinking of those turnarounds:
> 
> * As mentioned in this thread TC seems to be able to perform this task: https://www.ghisler.ch/board/viewtopic.php?t=39156
> * Copy the file with a script similar to the one proposed by skif-off, but with the ability to follow symlinks;
> * Copy the file from the left panel to all right tabs directories.
> 
> Do you think that this would be achievable ?
> 
> Thanks a lot for your attention !
> 
> Regards,
> 
> Rémi

Hello, I simply have the different situation with you, the script can work very well in my syslinks (i.e., both panels are local folders), but not working when one panel is local and another is basically a network folder (e.g., via the FTP plugin or the webdav plugin of TC)

**Reply by rmdug** [reply]
Thank you for your answers.

I apologize, my links were just Windows shortcuts and actually not symlinks.
I check this point first and give it a try again : )
